### ISSUE TITLE
Explicitly using `jax.lax.stop_gradient` in modules.

### ISSUE THREAD

**Issue by ASEM000** [patrick-kidger/equinox#318]
Hello Patrick,

AFAIK the equinox Jax transformation split/combine predicate is Jax arrays, not arraylike (i.e. excludes float and complex).

IMO, this asymmetry will create a silent problem when these modules are used with Jax transformation for modules with float/complex leaves. 

for example 

```python

layer = eqx.nn.Dropout(p=0.5, inference=None)
x = jnp.ones((10), dtype=jnp.float32)

@jax.grad
def loss_func(NN, x):
    return sum(NN(x, key=jax.random.PRNGKey(0)))

print(loss_func(layer, x).p) # -> `p` is trainable !
# 28.0

@eqx.filter_grad
def loss_func(NN, x):
    return sum(NN(x, key=jax.random.PRNGKey(0)))

print(loss_func(layer, x).p)
# None

```

I suggest explicitly using `jax.lax.stop_gradient` on non-trainable floats/complexes that are part of the module.

**Comment by patrick-kidger** [comment]
Agreed! This hadn't been a priority as the usual pattern with Equinox is to use the filtered transformations, but you're right that we might as well avoid a possible footgun here.

Have you identified any cases other than `dropout.p` that would need tackling?

**Comment by ASEM000** [comment]
I will post here if I find any other instances. 👍

**Comment by patrick-kidger** [comment]
Fixed in #319!

Let me know if you spot any other cases that need tackling like this.
